src/ssh/sftp_client.cpp

```diff
@@ -154,25 +142,43 @@ void mp::SFTPClient::pull_file(const std::string& source_path, const std::string
     } while (total < size);
 }
 
-void mp::SFTPClient::stream_file(const std::string& destination_path)
+void mp::SFTPClient::stream_file(const std::string& destination_path, std::istream& cin)
 {
     auto full_destination_path = full_destination(destination_path, stream_file_name);
-    SFTPSessionUPtr sftp{make_sftp_session(*ssh_session)};
-    SSH::throw_on_error(sftp, *ssh_session, "[sftp stream] init session failed", sftp_init);
-
     SFTPFileUPtr file_handle{
         sftp_open(sftp.get(), full_destination_path.c_str(), O_WRONLY | O_CREAT | O_TRUNC, file_mode), sftp_close};
     SSH::throw_on_error(sftp, *ssh_session, "[sftp stream] open failed", sftp_get_error);
 
-    QTextStream in_stream(stdin);
-    QString text;
-    QByteArray data;
+    std::array<char, max_transfer> data;
+    while (!cin.eof())
+    {
+        cin.read(data.data(), data.size());
+        sftp_write(file_handle.get(), data.data(), cin.gcount());
+        SSH::throw_on_error(sftp, *ssh_session, "[sftp push] remote write failed", sftp_get_error);
+    }
+}
+
+void mp::SFTPClient::stream_file(const std::string& source_path, std::ostream& cout)
+{
+    SFTPAttrUPtr file_attributes{sftp_stat(sftp.get(), source_path.c_str()), sftp_attributes_free};
+    SSH::throw_on_error(sftp, *ssh_session, "[sftp pull] getting stat failed", sftp_get_error);
+
+    const auto size{file_attributes->size};
+
+    SFTPFileUPtr file_handle{sftp_open(sftp.get(), source_path.c_str(), O_RDONLY, file_mode), sftp_close};
+    SSH::throw_on_error(sftp, *ssh_session, "[sftp pull] open failed", sftp_get_error);
+
+    auto total{0u};
+    std::array<char, max_transfer> data;
     do
     {
-        text = in_stream.read(package_size);
-        data = text.toUtf8();
+        auto r = sftp_read(file_handle.get(), data.data(), data.size());
+        SSH::throw_on_error(sftp, *ssh_session, "[sftp pull] read failed", sftp_get_error);
 
-        sftp_write(file_handle.get(), data.data(), data.size());
-        SSH::throw_on_error(sftp, *ssh_session, "[sftp push] remote write failed", sftp_get_error);
-    } while (!in_stream.atEnd());
+        if (r == 0)
```

Comment:
The doc for `sftp_read` says "Number of bytes written, < 0 on error with ssh and sftp error set [...] @see sftp_get_error()". So, should `r < 0` be checked? Or are we sure `SSH::throw_on_error` is enough?

On the other hand, I did not see mention of 0 read. Do we need the break? Admittedly I did not look at the implementation, but couldn't it be due to a spurious wake or something?

Reply:
Well, if the errors are not getting set as the docs say, then that is a bug in `libssh`.  So I think it's enough to check for the errors via `SSH::throw_on_error()`.

Regarding `sftp_read` returning 0, reading the code for `sftp_read`, it returns 0 when `eof` is reached.  Here's the particular code:
```
  if (handle->eof) {
    return 0;
  }
```

So I think it's safe to break here since we've reached the end of the file.

Reply:
OK then